torch/nn/parallel/comm.py

```diff
@@ -54,6 +54,7 @@ def broadcast_coalesced(tensors, devices, buffer_size=10485760):
         A tuple containing copies of :attr:`tensor`, placed on :attr:`devices`.
     """
     devices = [_get_device_index(d) for d in devices]
+    tensors = [torch.view_as_real(t) if t.is_complex() else t for t in tensors]
```

Comment:
Do you mind providing an example of the dimension mismatch you were seeing here?

Reply:
Running this test gets me:

```
Original Traceback (most recent call last):
  File "/data/users/sidneyfletcher/pytorch/torch/nn/parallel/parallel_apply.py", line 61, in _worker
    output = module(*input, **kwargs)
  File "/data/users/sidneyfletcher/pytorch/torch/nn/modules/module.py", line 744, in _call_impl
    result = self.forward(*input, **kwargs)
  File "test/distributed/test_data_parallel.py", line 610, in forward
    return x + self.cplx
RuntimeError: The size of tensor a (10) must match the size of tensor b (2) at non-singleton dimension 2
```

Reply:
I see. Since you've implemented complex tensor support in broadcast coalesced, tensors that are replicated by DataParallel (such as `cplx`) have been converted to reals - so the size of `cplx` changes from (1, 10) to (1, 10, 2). However, the input tensors, which are sent to each replica using the scatter collective, are still in complex mode, so their shape is still (1, 10). This line complains that addition of a (1, 10, 2)-shaped tensor and (1, 10)- shaped tensor is invalid.

The solution here would be to add complex-tensor support to the `scatter` collective.  Looking at the implementation of DataParallel's forward function, you may have to add support to the `gather` collective as well. I think the implementation could be similar to your current implementation for `broadcast_coalesced` in the corresponding functions in comm.py.

Reply:
Cool but what about the case where we want to multiply the matrices? (Ignoring that C and R^2 mult are not the same) - won't the dimensions still be wrong?

Reply:
I believe as long as the dimensions match for initial operation (defined by the user in the fwd pass), then this should work. We are only adding an extra dimension of size 2 to each tensor

fwiw, I tried the above example with multiplication and `x * cplx` and `torch.mul(s, clpx)` work fine

Reply:
ready for review if you get a chance?